src/sign/keyset.rs

```diff
@@ -0,0 +1,1906 @@
+
+        Ok(rolltype.roll_actions_fn()(next_state))
+    }
+
+    /// A report that the propagation of the second change has completed.
+    ///
+    /// The user reports that the changes have propagated and provides the
+    /// required ttl value. The actual changes to monitor were specified as on
+    /// action in the previous state (cache_expired1). A list of actions is
+    /// returned.
+    pub fn propagation2_complete(
+        &mut self,
+        rolltype: RollType,
+        ttl: u32,
+    ) -> Result<Vec<Action>, Error> {
+        // First check if the current roll state is Propagation2.
+        let Some(RollState::Propagation2) = self.rollstates.get(&rolltype)
+        else {
+            return Err(Error::WrongStateForRollOperation);
+        };
+        let next_state = RollState::CacheExpire2(ttl);
+        rolltype.rollfn()(RollOp::Propagation2, self)?;
+        self.rollstates.insert(rolltype.clone(), next_state.clone());
+        Ok(rolltype.roll_actions_fn()(next_state))
+    }
+
+    /// A report that any cached values should have expired by now.
+    ///
+    /// This method should be called at least ttl seconds after the call to
+    /// propagation2_complete where ttl is the value of the ttl parameter
+    /// passed to the propagation2_complete method. A list of actions is
+    /// returned.
+    pub fn cache_expired2(
+        &mut self,
+        rolltype: RollType,
+    ) -> Result<Vec<Action>, Error> {
+        // First check if the current roll state is CacheExpire2.
+        let Some(RollState::CacheExpire2(ttl)) =
+            self.rollstates.get(&rolltype)
+        else {
+            return Err(Error::WrongStateForRollOperation);
+        };
+        let next_state = RollState::Done;
+        rolltype.rollfn()(RollOp::CacheExpire2(*ttl), self)?;
+        self.rollstates.insert(rolltype.clone(), next_state.clone());
+
+        Ok(rolltype.roll_actions_fn()(next_state))
+    }
+
+    /// The user reports that all actions have been performed and that the
+    /// key roll is now complete.
+    pub fn roll_done(
+        &mut self,
+        rolltype: RollType,
+    ) -> Result<Vec<Action>, Error> {
+        // First check if the current roll state is Done.
+        let Some(RollState::Done) = self.rollstates.get(&rolltype) else {
+            return Err(Error::WrongStateForRollOperation);
+        };
+        rolltype.rollfn()(RollOp::Done, self)?;
+        self.rollstates.remove(&rolltype);
+        Ok(Vec::new())
+    }
+
+    /// Return the actions that need to be performed for the current
+    /// roll state.
+    pub fn actions(
+        &mut self,
+        rolltype: RollType,
+    ) -> Result<Vec<Action>, Error> {
+        if let Some(rollstate) = self.rollstates.get(&rolltype) {
+            Ok(rolltype.roll_actions_fn()(rollstate.clone()))
+        } else {
+            Ok(Vec::new())
+        }
+    }
+
+    fn update_ksk(
+        &mut self,
+        mode: Mode,
```

Comment:
I understand that the dry-run / real-run technique simplifies the code a bit, but I think it's also a bit confusing to use;  I think this function will become substantially simpler once `.keys` is a `HashMap`, and then the partial duplication necessary for error checking would be much more manageable.  Let's see how the code simplifies, and then perhaps we can consider it.